packages/fxa-auth-server/lib/payments/google-play/purchase-manager.ts

```diff
@@ -0,0 +1,334 @@
+/**
+ * Copyright 2018 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+import { CollectionReference } from '@google-cloud/firestore';
+import { androidpublisher_v3 } from 'googleapis';
+import Container from 'typedi';
+
+import { AuthLogger } from '../../types';
+import {
+  mergePurchaseWithFirestorePurchaseRecord,
+  SubscriptionPurchase,
+} from './subscription-purchase';
+import {
+  NotificationType,
+  Purchase,
+  PurchaseQueryError,
+  PurchaseUpdateError,
+  SkuType,
+} from './types';
+
+const REPLACED_PURCHASE_USERID_PLACEHOLDER = 'invalid';
+
+/*
+ * A class that provides user-purchase linking features
+ */
+export class PurchaseManager {
+  private log: AuthLogger;
+
+  /*
+   * This class is intended to be initialized by the library.
+   * Library consumer should not initialize this class themselves.
+   */
+  constructor(
+    private purchasesDbRef: CollectionReference,
+    private playDeveloperApiClient: androidpublisher_v3.Androidpublisher
+  ) {
+    this.log = Container.get(AuthLogger);
+  }
+
+  /*
+   * Query a subscription purchase by its package name, product Id (sku) and purchase token.
+   * The method queries Google Play Developer API to get the latest status of the purchase,
+   * then merge it with purchase ownership info stored in the library's managed Firestore database,
+   * then returns the merge information as a SubscriptionPurchase to its caller.
+   *  - triggerNotificationType is only necessary if the purchase query action is triggered by a Realtime Developer notification
+   */
+  public async querySubscriptionPurchase(
+    packageName: string,
+    sku: string,
+    purchaseToken: string,
+    triggerNotificationType?: NotificationType
+  ): Promise<SubscriptionPurchase> {
+    // STEP 1. Query Play Developer API to verify the purchase token
+    const apiResponse = await new Promise((resolve, reject) => {
+      this.playDeveloperApiClient.purchases.subscriptions.get(
```

Comment:
`googleapis` has a promise api too afaict

Reply:
It does, I copied the original sample code here, and was tempted to change it to a try/except, but once one factors in the `let` statement, capturing the exception, applying the transform, etc. it seemed about the same LoC so I left it as is.